include/mata/nfa/delta.hh

```diff
@@ -82,6 +81,11 @@ public:
     // but useful for adding states in a random order to sort later (supposedly more efficient than inserting in a random order)
     void inline push_back(const State s) { targets.push_back(s); }
 
+    template <typename... Args>
+    StateSet& emplace_back(Args&&... args) {
+        return targets.emplace_back(std::forward<Args>(args)...);
+    }
+
```

Comment:
What does this do? Or how is this thing called? Because for me this is C++ dark magic.

Reply:
It is a dark C++ magic, but in this case an understandable one still. It is called *variadic templates*, i.e., a template function with a variable number of arguments of the template type. We need this for the `emplace_back()` functions since the purpose of the `emplace_back()` is to accept a variable number of arguments, and try to construct an object of the type of the elements in the container with the passed arguments. That can work with zero, one, two or more arguments, each version constructing an object through possibly a different constructor of that type.

So we accept a variable number of arguments, but we do not want to construct the object in this function (in its scope) as the construction should only happen on the underlying vector `targets` in place. That means that we need to take these arguments without unpacking them or in any way resolving the pack of arguments, but directly pass them along (`std::forward` as the same pack of type `Args`) to another function taking a variadic pack of arguments, in this case the standard `std::vector<T>::emplace_back()` which has the same signature as this function, but inside the standard method there is already the allocation of the space in the underlying vector being performed, constructor of the underlying type of the elements of the vector is being called and so on. Only there will the pack be unpacked, and the arguments will be tried against all possible constructors for the underlying type of the elements as if you were calling the constructors manually as `TypeConstructor(arg1, arg2, ..., argn)`.

Reply:
I would put some comment here, something like "forwarding variadic templates to std::vector implementation".

Reply:
The comment has been added. It should be easier to look the relevant information up if necessary.